README.md
courier
=======

[![Build Status](https://travis-ci.org/hargettp/courier.svg?branch=master)](https://travis-ci.org/hargettp/courier)

A basic message-passing library, intended for simplifying network applications.

Inspired by Erlang's simple message-passing facilities, courier provides roughly similar capabilities.  Applications simply
create one or more endpoints, bind each to a transport using a given name, then can freely send / receive messages to
other endpoints just by referencing the name each endpoint bound to its transport.

A primary driver of courier's design is the ability of application writers to swap out the transport used for
delivering messages between endpoints without altering the logic of their application.  Many algorithms (such as for
distributed consensus) require a simplified message-passing facility in their implementation, and keeping the
transport implementation separate from the specific message passing protocol can simplify development and testing of
such algorithms.

The semantics of courier's use are simple:

 * An application sends and receives messages through an endpoint
 * Messages are just arbitrary bytestrings; application are free to construct / interpret them as needed
 * Endpoints are created with one or more transports
 * Sending messages is non-blocking and provides no feedback or guarantee regarding delivery; specific transports may,
   however, provide out of band feedback regarding delivery
 * Receiving messages is by default blocking, although a blocking call with a timeout is available

Both in-delivery-order reception and selective out-of-order reception are supported: applications designers
are free to choose which style makes sense for a specific application.  The use of the latter technique
approximates an Erlang-style coding model, as messages selected witha supplied function are received before
other non-selected messages that are ahead of them in the queue.

A sample of use follows:

```haskell
module Main where

 -- Import this package to manage endpoints
import Network.Endpoints

 -- A specific transport is necessary, however
import Network.Transport.Sockets.TCP

-- Needed for serialization
import Data.Serialize

main :: IO ()
main = do
  -- each endpoint needs a name; since we're using TCP
  -- as our transport, they need to be host/port pairs
  let name1 = Name "localhost:9001"
      name2 = Name "localhost:9002"
      -- the default resolvers just pull apart a name into separate
      -- host and port components; more elaborate resolvers could
      -- perform name lookups or other translations
      resolver = tcpSocketResolver4
  -- we need endpoints for each end of the communication
  endpoint1 <- newEndpoint
  endpoint2 <- newEndpoint
  -- we need a transport to move messages between endpoints
  withTransport (newTCPTransport4 resolver) $ \transport ->
    withEndpoint transport endpoint1 $
      withEndpoint transport endpoint2 $
      -- the first endpoint is just a client, so it needs a name to receive
      -- responses, but does not need a binding since it isn't accept connections
        withName endpoint1 name1 $
          -- the second endpoint is a server, so it needs a binding
          withBinding transport endpoint2 name2 $
            -- a connection between the first endpoint and the name of the second
            -- creates a bi-directional path for messages to flow between the endpoints
            withConnection transport endpoint1 name2 $ do
              sendMessage endpoint1 name2 $ encode "hello world!"
              msg <- receiveMessage endpoint2
              let Right txt = decode msg
                  in print (txt :: String) 
```

To install, simply run the following in a shell:

```
cabal install courier
```

Or if using cabal-dev, run in a directory prepared with a cabal-dev sandbox:

```
cabal-dev install courier
```

With the  advent of sandboxes in Cabal 1.18, cabal-dev is no longer necessary,
just use an updated cabal for simplicity. The example assumes you are building
courier from source:

```
cabal clone https://github.com/hargettp/courier.git
cd courier
cabal sandbox init
cabal install --only-dependencies
cabal install
```

Successfully installed on both Linux 12.04 AMD64 and Mac OS X Mount Lion , using GHC 7.6.3.

TODO

 * Ensure that new transports can be defined outside of the courier package and used with courier endpoints
 * Consider adding support for RAW sockets


changes.md
0.1.1.5

  * Added withClient / withServer helpers
  * Removed support for 7.6 / 7.8, as they are before the FAMP proposal was         
    implemented (https://wiki.haskell.org/Functor-Applicative-Monad_Proposal)
  * Added typed equivalents for a selection of functions in Network.RPC

0.1.1.4

  * Corrected an issue with memory transport where a message to an unbound destination would be dropped

0.1.1.3

  * Changed with... functions to return specified types in IO, instead of always returning (); not technically a
    breaking change, but does prevent backward compatibility from here
  * Memory transport flushes to destinations before completing shutdown
  * Linting and first compile against GHC 8.0.1; continuing to support back to 7.6 for now

0.1.1.2

  * Updated README with better working example
  * Removed warning about TCP/IP transport; it's working again

0.1.1.1

There are significant changes in this release, due to a complete rewrite of TCP connection handling, and,
in so doing, a rewrite of the basic model for how transports and endpoints interact. This rewrite fixes
bug #10.

Endpoints now simply have inbound and outbound mailboxes, and a set of names to which the endpoint is bound
via a transport. The intent is that endpoints are discrete entities orthogonal to the transport which moves
messages between them. Its the job of the transport to move messages from the outbound mailbox of one endpoint
to the inbound mailbox of the endpoint bound to the name specified in the envelope for a given message.

New helpers for managing the lifecycle of bindings and connections now exist, and also a helper for setting the name
of a client.

0.1.0.15

    * Initial implementation of IPv6 transports (TCP and UDP)
    * Small cleanup for better thread safety
    * Refactored socket transports for more code sharing
    * Added anyCall
    * Exported RequestId and mkRequestId
    * Incorporated changes from carletes for running tests in parallel by allowing
    each test to bind to their own set of ports; on machines with multiple CPUs the speedup
    has a noticeable impact on producitivity
    * Added methodSelector for applications that may want to customize RPC handling

0.1.0.14

    * Included test source into tarball

0.1.0.13

    * Explicitly removed support for 7.4; 7.6 now required.

0.1.0.12

    * Fixed a long-standing issue concerning inconsistent failures. Turns out, on operating
    systems with both IPv4 and IPv6 support, the TCP and UDP transports (which were really
    only intended to be IPv4 for now) were not properly finding IPv4 addresses--in some
    cases, IPv6 were being used instead, resulting in errors.  Both TCP and UDP now
    pass (IPv4 only, of course), and in the future, IPv6 versions of these transports
    will be added.

0.1.0.11

    * API breaking change: RPC functions no longer take / return instances of Serialize,
    but they take / return Message values instead.  The reason for this change is that
    while courier does use Serialize internally to simplify a variety of operations,
    putting instances of Serialize in RPC functions was misleading: applications would
    still need to take care to ensure that different RPC requests / responses could
    be differentiated, otherwise courier might accidentally deserialize a message
    to the wrong type at an application's request (e.g., inside a hear callback). This
    can lead to decoding errors, which may actually causes exceptions to be thrown
    in pure methods.  By leaving the interface to use Message values, its not only
    consistent with the rest of Endpoint behavior, but it clarifies the responsibility
    of the calling application to manage their Message payloads on their own.

    * Improvements to RPC. Previously, it was not possible to differentiate correctly
    between requests and responses in the same mailbox: the only check was that
    deserialization succeeded, which wasn't a sufficient test, and it would be possible
    to deserialize a response as a request, etc.  This lead to spurious errors.

    * Added functions for detecting presence of messages in an an endpoint's mailbox
    (while still leaving them unconsumed): mostly added this for applications that
    need to react to the presence of a particular message, but not actual consume
    such messages.

    * Exposed the Request / Response types used by RPC.  Applications may benefit
    from detecting them or otherwise manipulating them outside of normal RPC
    flow, they are exposed here for that purpose.

0.1.0.10

    * Added hear primitive as one-shot counterpart to call.  While handle
    may be generally useful for many applications as it fully encapsulates repeated
    handling of incoming requests, hear adds the ability to wait for a single
    incoming request. Hear also returns a function of type Reply suitable for sending
    a single response to the caller. Being more low-level in nature, some applications
    find direct use of hear more flexible than the higher-level handle construct.

    * Fixed a mistake in 0.1.0.9 where RPCs where supposed to have unique identifiers
    for matching up a call with a response.  The value was never computed, when it
    should have been a random value for every call, so that responses could be matched
    with their originating request.  Now using Data.UUID.V4 to generate unique identifiers
    for each request; dependency added in cabal.

    * Unit tests on OS X 10.9 unreliable again, with both ghc 7.6.3 and ghc 7.8.1 (rc1)

    * Unit tests passing on Ubuntu 12.04 LTS AMD64 ghc 7.6.3

    * Unit tests passing on FP Complete with 7.8 preview compiler

0.1.0.9

    * Added HelloWorld example, and updated code sample in documentation to reflect
      working with the current API

    * Early implementation of synchronous RPCs on top of endpoints, with basic unit tests.
      Both single and group RPCs are implemented, as are calls with definite and indefinite

src/Network/Endpoints.hs
{-# LANGUAGE DeriveDataTypeable     #-}
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE OverloadedStrings #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Network.Endpoints
-- Copyright   :  (c) Phil Hargett 2013
-- License     :  MIT (see LICENSE file)
--
-- Maintainer  :  phil@haphazardhouse.net
-- Stability   :  experimental
-- Portability :  non-portable (requires STM)
--
-- 'Endpoint's are a generalized abstraction for communication between parts of a program,
-- whether on the same physical host or distributed over a network. 'Endpoint's are intended
-- to simplify the development of network-centric applications by providing a small transport-independent
-- message-passing interface, and application writers can independently alter their implementation
-- by enabling their 'Endpoint's with different 'Network.Transport's without modifying the logic of their
-- application that sends / receives 'Message's.
--
-----------------------------------------------------------------------------

module Network.Endpoints (

  -- * How to use courier in an application
  -- $use

  -- * Primary API
  Endpoint(..),
  newEndpoint,

  withName,
  bindName,
  unbindName,
  BindException(..),

  sendMessage,
  broadcastMessage,
  receiveMessage,
  receiveMessageTimeout,
  postMessage,

  -- * Selective message reception
  selectMessage,
  selectMessageTimeout,
  detectMessage,
  detectMessageTimeout,

  -- * Other types
  Envelope(..),
  Message,
  Name(..),

  ) where

-- local imports

-- external imports

import Control.Concurrent
import Control.Concurrent.Async
import Control.Concurrent.Mailbox
import Control.Concurrent.STM
import Control.Exception

import qualified Data.ByteString as B
import Data.Serialize
import qualified Data.Set as S
import Data.Typeable

import GHC.Generics

--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

_log :: String
_log = "transport.endpoints"

-- $use
--
-- A sample of how to use this library:
--
-- > module Main where
-- >
-- >  -- Import this package to manage endpoints
-- > import Network.Endpoints
-- >
-- >  -- A specific transport is necessary, however
-- > import Network.Transport.Sockets.TCP
-- >
-- > -- Needed for serialization
-- > import Data.Serialize
-- >
-- > main :: IO ()
-- > main = do
-- >   -- each endpoint needs a name; since we're using TCP
-- >   -- as our transport, they need to be host/port pairs
-- >   let name1 = Name "localhost:9001"
-- >       name2 = Name "localhost:9002"
-- >       -- the default resolvers just pull apart a name into separate
-- >       -- host and port components; more elaborate resolvers could
-- >       -- perform name lookups or other translations
-- >       resolver = tcpSocketResolver4
-- >   -- we need endpoints for each end of the communication
-- >   endpoint1 <- newEndpoint
-- >   endpoint2 <- newEndpoint
-- >   -- we need a transport to move messages between endpoints
-- >   withTransport (newTCPTransport4 resolver) $ \transport ->
-- >     withEndpoint transport endpoint1 $
-- >       withEndpoint transport endpoint2 $
-- >       -- the first endpoint is just a client, so it needs a name to receive
-- >       -- responses, but does not need a binding since it isn't accept connections
-- >         withName endpoint1 name1 $
-- >           -- the second endpoint is a server, so it needs a binding
-- >           withBinding transport endpoint2 name2 $
-- >             -- a connection between the first endpoint and the name of the second
-- >             -- creates a bi-directional path for messages to flow between the endpoints
-- >             withConnection transport endpoint1 name2 $ do
-- >               sendMessage endpoint1 name2 $ encode "hello world!"
-- >               msg <- receiveMessage endpoint2
-- >               let Right txt = decode msg
-- >                   in print (txt :: String)
-- >
--

{-|
Endpoints are a locus of communication, used for sending and receive messages.
-}
data Endpoint = Endpoint {
  -- | The 'Transport' used by this 'Endpoint'
  -- endpointTransport :: Transport,
  -- | The 'Mailbox' where inbound 'Message's that the 'Endpoint' will receive are queued
  endpointInbound :: Mailbox Message,
  endpointOutbound :: Mailbox Envelope,
  -- | The 'Name's to which the 'Endpoint' is bound
  boundEndpointNames :: TVar (S.Set Name )
  }

{-|
Messages are containers for arbitrary data that may be sent to other 'Network.Endpoints.Endpoint's.
-}
type Message = B.ByteString

{-|
Name for uniquely identifying an 'Endpoint'; suitable for identifying
the target destination for a 'Message'. The specific interpretation of a name
is left to each 'Network.Transport.Transport'
-}
newtype Name = Name String deriving (Eq,Ord,Show,Generic)

instance Serialize Name

{-|
 An 'Envelope' wraps a 'Message' with the 'Name's of the destination for the message and (optionally)
the origin.
-}

data Envelope = Envelope {
  messageOrigin :: Maybe Name,
  messageDestination :: Name,
  envelopeMessage :: Message
} deriving (Eq,Show)

{-|
Create a new 'Endpoint' using the provided transports.
-}
newEndpoint :: IO Endpoint
newEndpoint = atomically $ do
  inbound <- newMailbox
  outbound <- newMailbox
  names <- newTVar S.empty
  return Endpoint {
    endpointInbound = inbound,
    endpointOutbound = outbound,
    boundEndpointNames = names
    }

{-|
Declare an 'Endpoint' as having the specified 'Name' while the supplied function executes. This can
often be useful for establishing the 'Name' of a client or initiator of a 'Network.Transport.Connection',
without requiring the client also have a 'Network.Transport.Binding'.

-}
withName :: Endpoint -> Name -> IO a -> IO a
withName endpoint origin actor = do
  atomically $ bindName endpoint origin
  finally actor $ atomically $ unbindName endpoint origin

{-|
Establish 'Name' as one of the 'boundEndpointNames' for an 'Endpoint'. Throws 'BindingExists' if
the 'Endpoint' is already bound to the 'Name'.
-}
bindName :: Endpoint -> Name -> STM ()
bindName endpoint name = do
  bindings <- readTVar $ boundEndpointNames endpoint
  if S.member name bindings
    then throw $ BindingExists name
    else modifyTVar (boundEndpointNames endpoint) $ S.insert name

{-|
Remove 'Name' as one of the 'boundEndpointNames' for an 'Endpoint'. Throws 'BindingDoesNotExist'
if the 'Endpoint' is not bound to the 'Name'.
-}
unbindName :: Endpoint -> Name -> STM ()
unbindName endpoint name = modifyTVar (boundEndpointNames endpoint) $ \bindings ->
  if S.member name bindings
    then S.delete name bindings
    else throw $ BindingDoesNotExist name

{-|
Exceptions generated when `Network.Transport.bind`ing a 'Name'.
-}
data BindException =
  BindingExists Name |
  BindingDoesNotExist Name
  deriving (Show,Typeable)

instance Exception BindException

{-|
Send a 'Message' to specific 'Name' via the indicated 'Endpoint'.
-}
sendMessage :: Endpoint -> Name -> Message -> IO ()
sendMessage endpoint name msg  = atomically $
    writeMailbox (endpointOutbound endpoint) $ Envelope Nothing name msg

{-|
Helper for sending a single 'Message' to several 'Endpoint's.
-}
broadcastMessage :: Endpoint -> [Name] -> Message -> IO ()
broadcastMessage endpoint names msg = mapM_ (\name -> sendMessage endpoint name msg) names

{-|
Receive the next 'Message' sent to the 'Endpoint', blocking until a message is available.
-}
receiveMessage :: Endpoint -> IO Message
receiveMessage endpoint = atomically $ readMailbox $ endpointInbound endpoint

{-|
Wait for a message to be received within the timeout, blocking until either a message
is available or the timeout has occurred.  If a message was available, returns @Just message@,
but returns @Nothing@ if no message available before the timeout occurred.
-}
receiveMessageTimeout :: Endpoint -> Int -> IO (Maybe Message)
receiveMessageTimeout endpoint delay = do
  resultOrTimeout <- race (receiveMessage endpoint) (threadDelay delay)
  case resultOrTimeout of
    Left result -> return $ Just result
    Right () -> return Nothing

{-|
Posts a 'Message' directly to an 'Endpoint', without use of a transport. This
may be useful for applications that prefer to use the 'Endpoint''s 'Mailbox'
as a general queue of ordered messages.
-}
postMessage :: Endpoint -> Message -> STM ()
postMessage endpoint = writeMailbox (endpointInbound endpoint)

{-|
Select the next available message in the 'Endpoint' 'Mailbox' matching
the supplied test function, or blocking until one is available. This function
differs from 'receiveMessage' in that it supports out of order message reception.
-}
selectMessage :: Endpoint -> (Message -> Maybe v) -> IO v
selectMessage endpoint testFn = atomically $ selectMailbox (endpointInbound endpoint) testFn

{-|
Wait for a message to be selected within the timeout, blocking until either a message
is available or the timeout has occurred.  If a message was available, returns @Just message@,
but returns @Nothing@ if no message available before the timeout occurred. Like
'selectMessage', this function enables out of order message reception.
-}
selectMessageTimeout :: Endpoint -> Int -> (Message -> Maybe v) -> IO (Maybe v)
selectMessageTimeout endpoint delay testFn = do
  resultOrTimeout <- race (selectMessage endpoint testFn) (threadDelay delay)
  case resultOrTimeout of
    Left result -> return $ Just result
    Right () -> return Nothing

{-|
Find a 'Message' in the 'Endpoint' 'Mailbox' matching the supplied
test function, or block until one is available.  Note that any such message
is left in the mailbox, and thus repeated calls to this function could find the
message if it is not consumed immediately.
-}
detectMessage :: Endpoint -> (Message -> Maybe v) -> IO v
detectMessage endpoint testFn = atomically $ findMailbox (endpointInbound endpoint) testFn

{-|
Find a 'Message' in the 'Endpoint' 'Mailbox' matching the supplied
test function, or block until either one is available or the timeout expires.
Note that any such message is left in the mailbox, and thus repeated calls
to this function could find the message if it is not consumed immediately.
-}
detectMessageTimeout :: Endpoint -> Int -> (Message -> Maybe v) -> IO (Maybe v)
detectMessageTimeout endpoint delay testFn = do
  resultOrTimeout <- race (detectMessage endpoint testFn) (threadDelay delay)
  case resultOrTimeout of
    Left result -> return $ Just result
    Right () -> return Nothing
